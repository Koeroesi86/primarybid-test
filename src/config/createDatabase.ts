import { Db, MongoClient } from 'mongodb';

let client: MongoClient;

const { DB_USERNAME, DB_PASSWORD, DB_HOST, DB_NAME } = process.env;

const createDatabase = async (): Promise<Db> => {
  if (!client) {
    client = new MongoClient(
      `mongodb://${DB_USERNAME}:${DB_PASSWORD}@${DB_HOST}?retryWrites=true&writeConcern=majority`
    );
    await client.connect();

    process.addListener('beforeExit', () => {
      if (client) {
        client.close();
      }
    });
  }

  return client.db(DB_NAME);
};

export default createDatabase;
