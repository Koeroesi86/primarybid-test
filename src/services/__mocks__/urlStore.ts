import ShortUrl from '../../types/ShortUrl';

export default {
  create: jest.fn(() => Promise.resolve({ token: '', url: '' } as ShortUrl)),
  getAll: jest.fn(() => Promise.resolve([] as ShortUrl[])),
  get: jest.fn(() => Promise.resolve({ token: '', url: '' } as ShortUrl)),
};
