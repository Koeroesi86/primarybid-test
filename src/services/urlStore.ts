import { Collection, Db } from 'mongodb';
import ShortUrl from '../types/ShortUrl';

const generateToken = (length = 8): string => {
  let chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
  return Array(length)
    .fill(1)
    .map(() => chars.charAt(Math.floor(Math.random() * chars.length)))
    .join('');
};

const ensureCollection = async (db: Db): Promise<Collection> => {
  if (!db.collection('urls')) await db.createCollection('urls');

  return db.collection('urls');
};

export default {
  create: async (db: Db, url: string): Promise<ShortUrl> => {
    if (!url) throw new Error('No url specified');

    const collection = await ensureCollection(db);
    await collection.createIndex('token');
    let iteration = 0;
    let token = generateToken();

    const existing = await collection.findOne({ url });
    if (existing) return { token: existing.token, url: existing.url };

    while (await collection.findOne({ token })) {
      iteration++;

      token = generateToken();

      if (iteration > 10) throw new Error('Too many retries');
    }

    await collection.insertMany([{ token, url }]);

    return collection.findOne({ token });
  },
  getAll: async (db: Db): Promise<ShortUrl[]> => {
    const collection = await ensureCollection(db);

    return (await collection.find().toArray()).map((saved) => ({
      token: saved.token,
      url: saved.url,
    }));
  },
  get: async (db: Db, token: string): Promise<ShortUrl> => {
    const collection = await ensureCollection(db);

    const saved = await collection.findOne({ token });
    if (!saved) throw new Error('Not found');

    return { token: saved.token, url: saved.url };
  },
};
