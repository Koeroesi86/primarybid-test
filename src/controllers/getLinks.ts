import express from 'express';
import { Db } from 'mongodb';
import urlStore from '../services/urlStore';
import ShortUrlResponse from '../types/ShortUrlResponse';

const getLinks = (db: Db) => async (
  req: express.Request,
  res: express.Response
): Promise<void> => {
  try {
    const saved = await urlStore.getAll(db);
    res.status(200);
    res.json(
      saved.map((short) => ({
        url: `${process.env.BASE_URL}/${short.token}`,
        original: short.url,
      })) as ShortUrlResponse[]
    );
  } catch (e) {
    res.status(400);
    res.json({});
  }
};

export default getLinks;
