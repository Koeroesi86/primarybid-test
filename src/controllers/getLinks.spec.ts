import { createRequest, createResponse } from 'node-mocks-http';
import getLinks from './getLinks';
import { mocked } from 'ts-jest/utils';
import urlStore from '../services/urlStore';
import createDatabase from '../config/createDatabase';

jest.mock('../services/urlStore');
jest.mock('mongodb');

describe('getLinks', () => {
  const mockedStore = mocked(urlStore, true);
  process.env.BASE_URL = 'https://test.com';

  it('should respond with success', async () => {
    const request = createRequest();
    const response = createResponse();
    const db = await createDatabase();
    mockedStore.getAll.mockResolvedValueOnce([
      { token: 'test1', url: 'https://example1.com' },
      { token: 'test2', url: 'https://example2.com' },
    ]);

    await getLinks(db)(request, response);

    expect(response.statusCode).toEqual(200);
    expect(response._getData()).toMatchSnapshot('data');
  });

  it('should respond error', async () => {
    const request = createRequest();
    const response = createResponse();
    const db = await createDatabase();
    mockedStore.getAll.mockRejectedValueOnce(new Error('test'));

    await getLinks(db)(request, response);

    expect(response.statusCode).toEqual(400);
    expect(response._getData()).toMatchSnapshot('data');
  });
});
