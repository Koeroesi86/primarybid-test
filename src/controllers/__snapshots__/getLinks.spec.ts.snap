// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getLinks should respond error: data 1`] = `"{}"`;

exports[`getLinks should respond with success: data 1`] = `"[{\\"url\\":\\"https://test.com/test1\\",\\"original\\":\\"https://example1.com\\"},{\\"url\\":\\"https://test.com/test2\\",\\"original\\":\\"https://example2.com\\"}]"`;
